Content.Server/Nuke/NukeSystem.Disk.cs

```diff
@@ -0,0 +1,87 @@
+﻿using Content.Server.Administration.Logs;
+using Content.Server.GameTicking;
+using Content.Server.Station.Components;
+using Content.Shared.Database;
+using Content.Shared.Maps;
+using Content.Shared.Nuke;
+using Robust.Shared.Map;
+
+namespace Content.Server.Nuke;
+
+public sealed partial class NukeSystem
+{
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly IAdminLogManager _adminLog = default!;
+
+    public void NukeDiskInitialize()
+    {
+        SubscribeLocalEvent<GameRunLevelChangedEvent>(OnRunLevelChangedDisk);
+        SubscribeLocalEvent<NukeDiskComponent, MapInitEvent>(OnMapDiskInit);
+        SubscribeLocalEvent<NukeDiskComponent, ComponentShutdown>(OnDiskShutdown);
+    }
+
+    private void OnRunLevelChangedDisk(GameRunLevelChangedEvent ev)
+    {
+
+        //Try to compensate for restartroundnow command
+        if (ev.Old == GameRunLevel.InRound && ev.New == GameRunLevel.PreRoundLobby)
+            OnRoundEnd();
+
+        switch (ev.New)
+        {
+            case GameRunLevel.PostRound:
+                OnRoundEnd();
+                break;
+        }
+    }
+
+    private void OnMapDiskInit(EntityUid uid, NukeDiskComponent component, MapInitEvent args)
+    {
+        var originStation = _stationSystem.GetOwningStation(uid);
+        var xform = Transform(uid);
+
+        if (originStation != null)
+        {
+            component.Station = originStation.Value;
+            component.StationMap = (xform.MapID, xform.GridUid);
+        }
+    }
+
+    private void OnDiskShutdown(EntityUid uid, NukeDiskComponent component, ComponentShutdown args)
+    {
+        var diskGridUid = component.StationMap.Item2;
+
+        if (!_mapManager.TryGetGrid(diskGridUid, out var grid) || !component.Respawn || !TryComp<StationMemberComponent>(diskGridUid, out var stationMember))
```

Comment:
This check will fail if the disk is in space or on a non-station grid, so need to change this if. The code later on handles it, but this code will exit early.

Reply:
Wouldn't it grab the grid uid from when it spawns initially? I'm not sure how it'd spawn in space or non station without admin spawning.

Reply:
Since it assigns station and grid on mapinit only, this shouldn't happen.